# CS 4661 Project: Spooky Author Identification (Logistic Regression Method)
### By Margarita Nuridjanian,

In [9]:
# Importing the required packages and libraries
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression

In [10]:
# Reading test CSV file directly from Web, and store it in a pandas DataFrame:
test_df=pd.read_csv("https://raw.githubusercontent.com/MargaritaNuridj/csula_data_science/master/spooky-author-identification/test.csv")

# Print first 5 rows
test_df.head()

id                                               text
0  id02310  Still, as I urged our leaving Ireland with suc...
1  id24541  If a fire wanted fanning, it could readily be ...
2  id00134  And when they had broken down the frail door t...
3  id27757  While I was thinking how I should possibly man...
4  id04081  I am not sure to what limit his knowledge may ...

In [13]:
# Reading train CSV file directly from Web, and store it in a pandas DataFrame:
train_df = pd.read_csv('https://raw.githubusercontent.com/MargaritaNuridj/csula_data_science/master/spooky-author-identification/train.csv')

# Print first 5 rows
train_df.head()
y_predict_lr = my_logreg.predict(X_tt_new)
print(y_predict_lr)

NameError: name 'my_logreg' is not defined

In [25]:
# Find unique value counts of each author
train_df['author'].value_counts()

# train_df = train_df[(train_df.author=='MWS') | (train_df.author=='EAP')]

EAP    7900
MWS    6044
Name: author, dtype: int64

In [28]:
# Split into features and labels
X_train = train_df['text']
y_train = train_df['author']
X_test = test_df['text']

X_train

0        This process, however, afforded me no means of...
2        In his left hand was a gold snuff box, from wh...
3        How lovely is spring As we looked from Windsor...
5        A youth passed in solitude, my best years spen...
6        The astronomer, perhaps, at this point, took r...
7              The surcingle hung in ribands from my body.
8        I knew that you could not say to yourself 'ste...
9        I confess that neither the structure of langua...
10       He shall find that I can feel my injuries; he ...
11       Here we barricaded ourselves, and, for the pre...
14        But a glance will show the fallacy of this idea.
15       He had escaped me, and I must commence a destr...
16       To these speeches they gave, of course, their ...
17       Her native sprightliness needed no undue excit...
18       I even went so far as to speak of a slightly h...
20       Now the net work was not permanently fastened ...
22       On every hand was a wilderness of balconies, o.

In [15]:
# Transform text data into vectors to run it through machine learning model
vectorizer = CountVectorizer(stop_words='english')
# vectorizer = CountVectorizer()

corpus = pd.concat([train_df['text'], test_df['text']])
vectorizer.fit(corpus)

CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words='english',
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)

In [16]:
# Randomly splitting the original dataset into training set and testing set
X_tr, X_tt, y_tr, y_tt = train_test_split(X_train, y_train, test_size=0.3, random_state=2)

X_tr_new = vectorizer.transform(X_tr)
X_tt_new = vectorizer.transform(X_tt)

In [17]:
# "my_logreg" is instantiated as an "object" of LogisticClassifier "class" 
my_logreg = LogisticRegression()
my_logreg.fit(X_tr_new, y_tr)

//anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
//anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [18]:
# Predicting the Binary Label:
y_predict_lr = my_logreg.predict(X_tt_new)
print(y_predict_lr)

['EAP' 'EAP' 'HPL' ... 'EAP' 'HPL' 'EAP']


In [19]:
# Estimating the probability (likelihood) of Each Label: 
y_pred_proba_lr = my_logreg.predict_proba(X_tt_new)
y_pred_proba_lr

array([[5.66797108e-01, 3.74438613e-01, 5.87642786e-02],
       [8.16034418e-01, 1.31032217e-01, 5.29333653e-02],
       [4.54477959e-02, 9.40162778e-01, 1.43894259e-02],
       ...,
       [5.38770971e-01, 2.81982504e-01, 1.79246524e-01],
       [1.05602369e-02, 9.89232189e-01, 2.07573860e-04],
       [4.92142249e-01, 1.11116964e-01, 3.96740787e-01]])

In [20]:
# Find accuracy score
score_lr = accuracy_score(y_tt, y_predict_lr)
print("Logistic Regression accuracy:", score_lr)

Logistic Regression accuracy: 0.7906026557711952


In [21]:
# Write data to CSV file
submission_lr = pd.DataFrame(y_pred_proba_lr, columns=["EAP","HPL","MWS"])
# submission_lr = pd.DataFrame(y_pred_proba_lr, columns=["EAP","MWS"])
submission_lr['id'] = test_df['id']
submission_lr = submission_lr[["id","EAP","HPL","MWS"]]
# submission_lr = submission_lr[["id","EAP","HPL","MWS"]]
submission_lr.to_csv('submission_lr.csv', index=None)

# Notes:
## With stop word and no filter accuracy is 0.7906026557711952
## If we do not include stop word accuracy is 0.8255022131426626
## If we do not include stop word and HPL the accuracy is 0.8697418738049714

In [22]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score




# Reading test CSV file directly from Web, and store it in a pandas DataFrame:
test_df=pd.read_csv("https://raw.githubusercontent.com/MargaritaNuridj/csula_data_science/master/spooky-author-identification/test.csv")

# Print first 5 rows
test_df.head()
# Reading train CSV file directly from Web, and store it in a pandas DataFrame:
train_df = pd.read_csv('https://raw.githubusercontent.com/MargaritaNuridj/csula_data_science/master/spooky-author-identification/train.csv')

# Print first 5 rows
train_df.head()
X_train = train_df['text']
y_train = train_df['author']
X_test = test_df['text']
vectorizer = CountVectorizer()

corpus = pd.concat([train_df['text'], test_df['text']])
vectorizer.fit(corpus)
X_tr, X_tt, y_tr, y_tt = train_test_split(X_train, y_train, test_size=0.3, random_state=2)

X_tr_new = vectorizer.transform(X_tr)
X_tt_new = vectorizer.transform(X_tt)
# "my_decisiontree" is instantiated as an "object" of LogisticClassifier "class" 
my_logreg = LogisticRegression()
my_logreg.fit(X_tr_new, y_tr)
# Predicting the Binary Label:
y_predict_lr = my_logreg.predict(X_tt_new)
# print(y_predict_lr)
# Find accuracy score
score_lr = accuracy_score(y_tt, y_predict_lr)
print("Logistic Regression accuracy:", score_lr)

//anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
//anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


Logistic Regression accuracy: 0.8255022131426626


In [23]:
# Reading test CSV file directly from Web, and store it in a pandas DataFrame:
test_df=pd.read_csv("https://raw.githubusercontent.com/MargaritaNuridj/csula_data_science/master/spooky-author-identification/test.csv")

# Print first 5 rows
test_df.head()
# Reading train CSV file directly from Web, and store it in a pandas DataFrame:
train_df = pd.read_csv('https://raw.githubusercontent.com/MargaritaNuridj/csula_data_science/master/spooky-author-identification/train.csv')

# Print first 5 rows
train_df.head()
train_df = train_df[(train_df.author=='MWS') | (train_df.author=='EAP')]
X_train = train_df['text']
y_train = train_df['author']
X_test = test_df['text']
vectorizer = CountVectorizer()

corpus = pd.concat([train_df['text'], test_df['text']])
vectorizer.fit(corpus)
X_tr, X_tt, y_tr, y_tt = train_test_split(X_train, y_train, test_size=0.3, random_state=2)

X_tr_new = vectorizer.transform(X_tr)
X_tt_new = vectorizer.transform(X_tt)
# "my_decisiontree" is instantiated as an "object" of LogisticClassifier "class" 
my_logreg = LogisticRegression()
my_logreg.fit(X_tr_new, y_tr)
# Predicting the Binary Label:
y_predict_lr = my_logreg.predict(X_tt_new)
# print(y_predict_lr)
# Find accuracy score
score_lr = accuracy_score(y_tt, y_predict_lr)
print("Logistic Regression accuracy:", score_lr)

Logistic Regression accuracy: 0.8697418738049714


//anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
